### ISSUE TITLE
[BUG] Errors with Chrome Driver 114

### ISSUE THREAD

**Issue by shopbuy11** [charlesbel/Microsoft-Rewards-Farmer#180]
The bot had worked fine for a few months without issue.  Then it stopped today when I updated to Chrome Driver 114, since Chrome auto-updated to that version.    Here is the error I got.   I have tried to roll back to Chrome 113 but same error persists.   Any idea would be greatly appreciated.

**Screenshot of the page**
![image](https://github.com/charlesbel/Microsoft-Rewards-Farmer/assets/136040705/d36bea0e-3fa8-4c3e-8e54-1464c742cb0d)

**Comment by lollogiaco** [comment]
delete chrome and install chromium 113, i did thi ìs but it gives some errors : DO NOT ATTEMPT TO RE-DO THE SCRIPT MULTIPLE TIMES (all my 5 accounts got banned)

**Comment by shopbuy11** [comment]
Yes, I updated my Chrome Driver to 114, but that didn't help.  Again, I even rolled back Chrome to 113 and reinstall Chrome Driver 113, but the error persisted anyway.   Don't know what caused it exactly...

**Comment by shopbuy11** [comment]
> chromium

Where do you get chromium 113?  Or do you mean Chrome 113?  I downloaded Chrome 113 from Slimjet, but again, that didn't solve my problem.

**Comment by shopbuy11** [comment]
I just noticed my MS Reward account got suspended, which explains why the bot no longer works. I will contact them and see if I can get a response as to why.

**Comment by alessiouw** [comment]
The bot works fine, they just banned your public IP (or maybe your hw info, like the NIC MAC address)

**Comment by shopbuy11** [comment]
Not really, as I was able to create 2 new accounts and farm on those.   I guess time will tell if they get banned too, but at least it's not immediate based on IP.

**Comment by shopbuy11** [comment]
I think there is still a bug in the code.   Even when my accounts are fine, the same error will still occur randomly.   When I keep retrying, eventually it will run through OK.   If anyone has ideas, they would be appreciated.   Again, it started with Chrome driver 114...

**Comment by shopbuy11** [comment]
The issue appears to be with getDashboardData function/call.  I am not well versed enough to figure it out.  But when I changed the code to skip this call, then it completes without error.  Unfortunately that means I can only get points for searches, which is better than nothing of course but the CompleteDailySet, CompletePunchCards and MorePromos all fail.   Maybe it's a MS change that broke the code...   Anyone any ideas on how to fix this call?

![image](https://github.com/charlesbel/Microsoft-Rewards-Farmer/assets/136040705/c92a39ce-2abd-4652-b036-777fa1b0b2c1)

**Comment by EvanzEssence** [comment]
I am getting the same error on both my Windows and Linux machines.  Im guessing it was a MS change?

**Comment by shopbuy11** [comment]
I am not good enough at python to fix the code in GetDashboardData, but I did figure out a way to get it to run without error, and at least completing the daily searches.   Not getting credit for the quizzes is fine for now.   Look for the "for account in ACCOUNTS" code in the end of the .py file and change the section highlighted to what's shown.

![image](https://github.com/charlesbel/Microsoft-Rewards-Farmer/assets/136040705/473bb02c-0fcf-46d9-8e51-e450c639d1ad)

Here is the code:

    import random
    from contextlib import suppress
    #remainingSearches, remainingSearchesM = getRemainingSearches(browser)
    #if remainingSearches != 0:
    print('[BING]', 'Starting Desktop and Edge Bing searches...')
    #bingSearches(browser, remainingSearches)
    bingSearches(browser, random.randint(30,35))
    prGreen('[BING] Finished Desktop and Edge Bing searches !')

    with suppress(Exception):
        print('[DAILY SET]', 'Trying to complete the Daily Set...')
        completeDailySet(browser)
        prGreen('[DAILY SET] Completed the Daily Set successfully !')
        print('[PUNCH CARDS]', 'Trying to complete the Punch Cards...')
        completePunchCards(browser)
        prGreen('[PUNCH CARDS] Completed the Punch Cards successfully !')
        print('[MORE PROMO]', 'Trying to complete More Promotions...')
        completeMorePromotions(browser)
        prGreen('[MORE PROMO] Completed More Promotions successfully !')
        
    browser.quit()

    #if remainingSearchesM != 0:
    browser = browserSetup(True, MOBILE_USER_AGENT)
    print('[LOGIN]', 'Logging-in...')
    login(browser, account['username'], account['password'], True)
    print('[LOGIN]', 'Logged-in successfully !')
    print('[BING]', 'Starting Mobile Bing searches...')
    bingSearches(browser, random.randint(20,25), True)
    prGreen('[BING] Finished Mobile Bing searches !')
    browser.quit()

**Comment by Bibekdka** [comment]
hi bro
are u using the microsoft farmer i tried it in my pc it didnt work can we
talk / discuss on it if u dont mind...
with regards
Bibek deka

**Comment by EvanzEssence** [comment]
> I am not good enough at python to fix the code in GetDashboardData, but I did figure out a way to move things around to get it to run without error, and at least completing the daily searches. Not getting credit for the quizzes is fine for now. Look for the "for account in ACCOUNTS" code in the end of the .py file and change the section highlighted to what's shown. Hope this helps.
> 
> ![image](https://user-images.githubusercontent.com/136040705/250680414-473bb02c-0fcf-46d9-8e51-e450c639d1ad.png)
> 
> Here is the code:
> 
> ```
> import random
> from contextlib import suppress
> #remainingSearches, remainingSearchesM = getRemainingSearches(browser)
> #if remainingSearches != 0:
> print('[BING]', 'Starting Desktop and Edge Bing searches...')
> #bingSearches(browser, remainingSearches)
> bingSearches(browser, random.randint(30,35))
> prGreen('[BING] Finished Desktop and Edge Bing searches !')
> 
> with suppress(Exception):
>     print('[DAILY SET]', 'Trying to complete the Daily Set...')
>     completeDailySet(browser)
>     prGreen('[DAILY SET] Completed the Daily Set successfully !')
>     print('[PUNCH CARDS]', 'Trying to complete the Punch Cards...')
>     completePunchCards(browser)
>     prGreen('[PUNCH CARDS] Completed the Punch Cards successfully !')
>     print('[MORE PROMO]', 'Trying to complete More Promotions...')
>     completeMorePromotions(browser)
>     prGreen('[MORE PROMO] Completed More Promotions successfully !')
>     
> browser.quit()
> 
> #if remainingSearchesM != 0:
> browser = browserSetup(True, MOBILE_USER_AGENT)
> print('[LOGIN]', 'Logging-in...')
> login(browser, account['username'], account['password'], True)
> print('[LOGIN]', 'Logged-in successfully !')
> print('[BING]', 'Starting Mobile Bing searches...')
> bingSearches(browser, random.randint(20,25), True)
> prGreen('[BING] Finished Mobile Bing searches !')
> browser.quit()
> ```

Awesome, thanks so much!  This works.  I was able to get this to run after playing with the indentation a little bit (I think it was a copy-paste issue).  I suggest people paste it into Notepad++ and see what it barks about

**Comment by jeannly** [comment]
> I am not good enough at python to fix the code in GetDashboardData, but I did figure out a way to move things around to get it to run without error, and at least completing the daily searches. Not getting credit for the quizzes is fine for now. Look for the "for account in ACCOUNTS" code in the end of the .py file and change the section highlighted to what's shown. Hope this helps.


Think I managed to figure out being able to get the points for cards again.
The `getDashboardData` function is broken due to some funny/broken Javascript being on Microsoft's website.

---
**I used your code + made the following changes:**
1. Replace the `getDashboardData` function definition (you can find the location in the code [here](https://github.com/charlesbel/Microsoft-Rewards-Farmer/blob/30c26d30ef0730183fe8bbda6ba24c1371b05e33/ms_rewards_farmer.py#L477-L480)):

**Original/Broken**:
```
def getDashboardData(browser: WebDriver) -> dict:
    dashboard = findBetween(browser.find_element(By.XPATH, '/html/body').get_attribute('innerHTML'), "var dashboard = ", ";\n        appDataModule.constant(\"prefetchedDashboard\", dashboard);")
    dashboard = json.loads(dashboard)
    return dashboard
```
**New**:
```
def getDashboardData(browser: WebDriver) -> dict:
    dashboard_from_window = WebDriverWait(browser, timeout=40).until(lambda browser: browser.execute_script("return window.dashboard;"))
    return dashboard_from_window
```
2. Comment out [ this else statement](https://github.com/charlesbel/Microsoft-Rewards-Farmer/blob/30c26d30ef0730183fe8bbda6ba24c1371b05e33/ms_rewards_farmer.py#L789-L791)

Original:
```
    if browser.find_element(By.XPATH, '//*[@id="navs"]/div/div/div/div/div[4]/a').get_attribute('target') == '_blank':
        BASE_URL = 'https://rewards.microsoft.com'
        browser.find_element(By.XPATH, '//*[@id="navs"]/div/div/div/div/div[4]/a').click()
        time.sleep(1)
        browser.switch_to.window(window_name=browser.window_handles[0])
        browser.close()
        browser.switch_to.window(window_name=browser.window_handles[0])
        time.sleep(10)
    else:
        BASE_URL = 'https://account.microsoft.com/rewards'
        browser.get(BASE_URL)
```
New:
```
    if browser.find_element(By.XPATH, '//*[@id="navs"]/div/div/div/div/div[4]/a').get_attribute('target') == '_blank':
        BASE_URL = 'https://rewards.microsoft.com'
        browser.find_element(By.XPATH, '//*[@id="navs"]/div/div/div/div/div[4]/a').click()
        time.sleep(1)
        browser.switch_to.window(window_name=browser.window_handles[0])
        browser.close()
        browser.switch_to.window(window_name=browser.window_handles[0])
        time.sleep(10)
    BASE_URL = 'https://account.microsoft.com/rewards'
    browser.get(BASE_URL)
```

it's working for me now, let me know if it works for anyone else

**Comment by charlesbel** [comment]
Fixed with v3.0 #188
